aptos-labs/aptos-core
# [testsuite][gha] use poetry everywhere and refactor testsuite

## Description

Attempts to unify all python usage across the entire repo. In general, this means that poetry is our entrypoint for all python utilities. This can generally be split into the categories of:
* test utilities in `testsuite`
* test utilities elsewhere (e.g. python SDK examples, aptos CLI and faucet integration tests, etc)
* extra scripts (e.g. in `scripts/` and elsewhere -- not supported)

### Summary of changes
#### Github Actions
* new `python-setup` action to unify and de-dup all places where we're installing python and basic deps
* pin `snok/install-poetry` to v1.3.3 by git commit hash
* use poetry everywhere we're calling python from an action

#### Testsuites
* New `Filesystem.exists(file_name)` method to test file existence
* `forge.py` no longer has to be run from repo root
* Make pyright happy again. Add types and assertions
* Remove deprecated Forge link shortening logic

#### Test framework
* Create a new testsuite entrypoint `testrun`, to do common things like calling `poetry` the right way and project path discovery. This gets rid of all the shell poetry wrappers (introduced https://github.com/aptos-labs/aptos-core/pull/7785) that start add a bit of clutter at scale
* Rename `forge_wrapper_core` to `test_framework`, as we have many python test utilities now that reuse the same framework
* Remove the `logger` decorator to make it play nice with pyright static checking. Simplify and just use a global logger definition

## How to land

Changes to test utilities and these actions should be landed against all supported branches, to ensure CI continues working as intended.

## Test Plan

Unit tests locally and in CI (using the updated actions). Tests have been written for all new test framework components, and some existing core framework stuff has new tests as well.

Some tests are expected to fail until this PR is landed in the `main` branch:
* Continuous Forge Tests - GCP / determine-test-metadata (we should just remove this test now that GCP forge is stable for 1+ weeks)
